cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java

```diff
@@ -463,14 +501,17 @@ private LocalFile uploadFile(String bucket, String targetFilePath,
       }
 
       Blob blob = storage.get(blobId);
-      if (!isCacheable) {
+      if (!isCacheable && !isCacheablePerVersion) {
         // Precondition fails means the blob already exists, most likely happens due to retries
         // https://cloud.google.com/storage/docs/request-preconditions#special-case
         // Overwrite the file
         BlobInfo existingBlobInfo = BlobInfo.newBuilder(blob.getBlobId()).setContentType(contentType).build();
         uploadToGCS(localFile.getURI(), storage, existingBlobInfo, Storage.BlobWriteOption.generationMatch());
         LOG.debug("Successfully uploaded file {} to gs://{}/{} by overwriting due to conflict",
                   localFile.getURI(), bucket, targetFilePath);
+      } else if (isCacheablePerVersion) {
```

Comment:
nit: Let's put log.debug in the else block. We can then remove `        DataprocUtils.setTemporaryHoldOnGCSObject(storage, bucket, blob, targetFilePath);` out of else block into a separate if block. This might be a bit cleaner.

Reply:
ack